### DISCUSSION TITLE
Animatable avatars for MuJoCo models in Unity

### DISCUSSION THREAD

**Discussion by Balint-H** [google-deepmind/mujoco#535] [Show and tell]
To be able to apply animations in the Unity plugin, or exploit all the nice rigging methods implemented in the engine for kinematic controllers, a model needs to be imported in an FBX format. To have a 1:1 equivalence with the MuJoCo model, it is important that they have matching local positions of body parts. To achieve this you can remove all Mj components from your humanoid model, and can then use the FBX exporter package in unity to create an animatable rig, which you can reimport. 

During the import process, if its a humanoid rig, you can select it as such. In this case you will probably have to edit the skeleton, and reassign a few of the transforms to the correct body-parts. Once done, you can track the individual transforms with MjMocapBodies to get their information to the simulation (or to connect regular bodies to them with constraints).

Here's an example I've been working with for the humanoid model from ([Merel et al. 2017](https://arxiv.org/abs/1707.02201)):
[humanoid_CMU.zip](https://github.com/deepmind/mujoco/files/9808901/humanoid_CMU.zip)

If someone has a recommendation on how to combine the individual meshes into a single weighted mesh for neater material assignment that doesn't involve manually recreating the skeleton and painting the vertex weights, let me know. For the time being I've been using an editor script to apply a material to all meshrenderers of the animated model.

It's also possible to do it with the traditional MuJoCo humanoid (e.g. [Tassa et al. 2012](https://ieeexplore.ieee.org/abstract/document/6386025)), but for humanoid rigs in Unity it is important to have the character in a T-pose. To achieve this, you can modify the local transform of body segments, but don't offset the Geom objects themselves.

**Comment by shubhankarKulk** [comment]
Can we convert fbx animations so that they can be replayed on humanoid models in MuJoCo?

**Reply by Balint-H** [reply]
With the Unity plugin the above method is a way to do it (with animated mocap bodies). For classic MuJoCo: if there's mismatch between the animation and humanoid, you need to retarget the animation, which might involve inverse kinematics.  The dm_control project has some examples of converting animations to joint angles in mujoco models:
https://github.com/google-deepmind/dm_control/blob/46390cfc356dfcb4235a2417efb2c3ab260194b8/dm_control/suite/utils/parse_amc.py#L27

**Reply by shubhankarKulk** [reply]
This utilizes a given environment for its physics. I was wondering if it could be possible to run any mixamo animations in MuJoCo. I have successfully converted the file into readable data and am now doing some preprocessing so that I can run it on the model, but I have come to realise that the root joint for the animation and the humanoid is completely different, i.e. the root body for the humanoid is the torso and the root body for the animation are the hips. Is there a way around it?

**Reply by Balint-H** [reply]
Mismatches between animation hierarchies is why you'd need retargeting and IK. The specific script from dm_control I sent doesnt actually use physics (if I remember it correctly), rather just converts kinematics (animation) to the state space of the MuJoCo model so it can be played back.

For your specific problem a manual solution could be adding the missing dofs and parent root body to the mj model, or processing the animation so the animation root is the pelvis.

**Reply by shubhankarKulk** [reply]
I created a model which is has a root body on the pelvis and added the rest of the limbs accordingly. Additionally, I am trying to get how to convert the fbx motion capture data to a mujoco humanoid, but I am encountering problems with the rotations. When I load the fbx data in blender, there is a 90 degree rotation on the x axis. I tried to correct this rotation in mujoco, but have been unsuccessful. Any help would be appreciated. The animations I am trying to use are default mixamo fbx animation, as you can see in the actualMotionCapture video, and the improvedImplementation video is the one I have tried to implement.
[actualMotionCapture.webm](https://github.com/user-attachments/assets/e1450df1-b72c-4fb3-aec1-122d06922099)
[improvedImplementation.webm](https://github.com/user-attachments/assets/99d86be7-b7b6-45b8-9ad4-893c7e27f79d)

**Reply by Balint-H** [reply]
Converting motion file formats to one another can be really tricky, I've banged my head against similar problems in the past 😄 

Once you load the anim in blender how do you continue to process it?
I've found that BVH is a bit easier to interpret than FBX formats, so you may want to export to that and work with its serial euler angles format.

**Reply by shubhankarKulk** [reply]
That is interesting. I usually converted the fbx format to a text document containing positions and orientations (quaternions). I simply remapped these values to the qpos values of the humanoid that I created. It gives somewhat of a promising result. I will have a look into bvh file format, as I believe there is a blender script available for direct conversion of fbx to bvh file formats. I am also unable to understand how bvh files are structured. Could you help me out in this regard?

**Reply by Balint-H** [reply]
The BVH exporter is built into blender which is quite convenient. Just select your animated rig and export it under File. Here's a nice breakdown on BVH format: https://staffwww.dcs.shef.ac.uk/people/S.Maddock/publications/Motion%20Capture%20File%20Formats%20Explained.pdf
